startup/81-pe-callbacks.py

```diff
@@ -68,11 +108,123 @@ def event_page(self, doc):
             return doc
 
 
-# Background subtraction callback
-# TODO: this needs to be a wrapped RunRouter, figure out later
-bgsub_callback = DarkSubtractionCallback(image_key="pe1_image",
-                                         primary_stream="primary",
-                                         dark_stream="dark")
+# From Tom on 02/21/2019:
+class DarkFrameCache(Device):
+    def __init__(self, *args, **kwargs):
+        # self.det = det
+        self.last_collected = None
+        self.just_started = True
+        self.update_done = False
+        return super().__init__(*args, **kwargs)
+
+    def read(self):
+        return self._read
+
+    def read_configuration(self):
+        return self._read_configuration
+
+    @property
+    def configuration_attrs(self):
+        return self._configuration_attrs
+
+    @property
+    def read_attrs(self):
+        return self._read_attrs
+
+    def describe(self):
+        return self._describe
+
+    def describe_configuration(self):
+        return self._describe_configuration
+
+    # def describe_configuration(self):
+    #     return self.det.describe_configuration
+
+    def collect_asset_docs(self):
+        # keep track of when we get restaged to restore these
+        yield from self._asset_docs_cache
+        self._really_cached = self._asset_docs_cache
+        self._asset_docs_cache = []
+
+    def stage(self):
+        self._asset_docs_cache = self._really_cached
+
+
+def dark_plan(cam, dark_frame_cache, obsolete_secs, shutter):
+    if (dark_frame_cache.just_started or  # first run after instantiation
+        (dark_frame_cache.last_collected is not None and
+         time.monotonic() - dark_frame_cache.last_collected > obsolete_secs)):
+        init_shutter_state = shutter.status.get()
```

Comment:
better do do this with a `read` (but is very verbose `init_shutter_state = (yield from bs.read(shutter))[shutter,name]['value'])`)

Reply:
but that does not actually work as written because read returns None if you are listifying the plan.

Reply:
Sorry, I am a bit confused. Is it OK to keep the existing `shutter.status.get()` here?

Reply:
It is read-only so it is not _dangerous_, but it is less than great.  plans really should not be directly touching the underlying hardware.  I think:

```python
tmp = yield from bps.read(shutter.status)
init_shutter_state = tmp[shutter.status.name]['value'] if tmp is not None else None
```

will actually work (we really should put that in a helper function!).

Reply:
Nice, I like it. This should work fine with the simulators without touching the hardware.